cinder/volume/drivers/nexenta/ns5/nfs.py

```diff
@@ -237,24 +264,114 @@ def migrate_volume(self, ctxt, volume, host):
+                greenthread.sleep(timeout)
+                timeout = timeout * 2
+            else:
+                url = 'hpr/services/%s%s' % (svc_name, params)
+                self.nef.delete(url)
+                return false_ret
+
+        url = 'hpr/services/%s%s' % (svc_name, params)
         self.nef.delete(url)
+
+        try:
+            self.delete_volume(volume)
+        except exception.NexentaException as exc:
+            LOG.warning("Cannot delete source volume %(volume)s on "
+                        "NexentaStor Appliance: %(exc)s",
+                        {'volume': volume['name'], 'exc': exc})
+
         return True, {'provider_location': provider_location}
 
+    def initialize_connection(self, volume, connector):
+        LOG.debug('Initialize volume connection for %s', volume['name'])
+        url = 'hpr/activate'
+        data = {'datasetName': volume['provider_location'].split(':/')[1]}
+        self.nef.post(url, data)
+        data = {'export': volume['provider_location'], 'name': 'volume'}
+        return {
+            'driver_volume_type': self.driver_volume_type,
+            'data': data,
+            'mount_point_base': self.nfs_mount_point_base
+        }
+
+    def retype(self, context, volume, new_type, diff, host):
+        """Convert the volume to be of the new type.
+
+        :param ctxt: Context
+        :param volume: A dictionary describing the volume to migrate
+        :param new_type: A dictionary describing the volume type to convert to
+        :param diff: A dictionary with the difference between the two types
+        :param host: A dictionary describing the host to migrate to, where
+                     host['host'] is its name, and host['capabilities'] is a
+                     dictionary of its reported capabilities.
+        """
+        LOG.debug('Retype volume request %(vol)s to be %(type)s '
+                  '(host: %(host)s), diff %(diff)s.',
+                  {'vol': volume['name'],
+                   'type': new_type,
+                   'host': host,
+                   'diff': diff})
+
+        retyped = False
+        migrated = False
+        model_update = None
+
+        src_driver = self.__class__.__name__
+        dst_driver = host['capabilities']['location_info'].split(':')[0]
+        if src_driver != dst_driver:
+            LOG.warning('Cannot retype from %(src_driver)s to '
+                        '%(dst_driver)s.',
+                        {
+                            'src_driver': src_driver,
+                            'dst_driver': dst_driver
+                        })
+            return False
+
+        old, new = (volume['host'], host['host'])
+        if old != new:
+            migrated, provider_location = self.migrate_volume(
+                context, volume, host)
+
+        if not migrated:
+            model_update = {'provider_location': volume['provider_location']}
+        return retyped or migrated, model_update
```

Comment:
retyped is never changed and is always False

Reply:
Retyped is still not fully functional and this is more like a placeholder for future

Reply:
retype is still not fully functional, so this is a placeholder for a later patch where "retyped' will change it's value. Do you think changing it to False for now will be a good idea or better leave it as is?

Reply:
It is up to you. As for me, I do not think adding an extra unused variable to the code just for possible future use is good. Because at this moment it is unclear how the code for retype will look like. E.g, what happens to lines 371-372 after adding retype? So, it may result in two separate branches for retype & migrate making both variables useless.

Reply:
actually it makes sense to remove retype completely from stable branches for now, you are right.